src/pages/plugin-trends/index.tsx

```diff
@@ -0,0 +1,131 @@
+// PluginTrends.tsx
+import { useState, useEffect } from 'react'
+import axios from 'axios'
+import {
+    Stack,
+    Pagination,
+    CircularProgress,
+    TextField,
+    Box,
+    Grid,
+    FormControl,
+    InputLabel,
+    Select,
+    MenuItem,
+} from '@mui/material'
+import NavBar from '../../components/NavBar'
+import { pluginList, IPluginData } from '../../data/plugins'
+import useSortPlugins from '../../hooks/useSortPlugins'
+import usePagination from '../../hooks/usePagination'
+import PluginCard from '../../components/PluginCard'
+
+const PluginTrends: React.FC = () => {
+    const [plugins, setPlugins] = useState<IPluginData[]>([])
+    const [filteredPlugins, setFilteredPlugins] = useState<IPluginData[]>([])
+    const [loading, setLoading] = useState<boolean>(true)
+    const [searchTerm, setSearchTerm] = useState<string>('')
+    const { sortOption, setSortOption } = useSortPlugins(filteredPlugins)
+    const itemsPerPage = 12
+
+    const { page, handlePageChange, paginatedData, totalPages } = usePagination(filteredPlugins, itemsPerPage)
+
+    useEffect(() => {
+        const fetchPluginList = async () => {
+            try {
+                const pluginDataPromises = pluginList.map((plugin) => {
+                    const url = `https://raw.githubusercontent.com/jenkins-infra/infra-statistics/gh-pages/plugin-installation-trend/${plugin.id}.stats.json`
+                    // console.log('Fetching URL:', url)
+                    return axios.get(url).then((response) => ({ ...plugin, chartData: response.data }))
+                })
+                const pluginData = await Promise.all(pluginDataPromises)
+                setPlugins(pluginData)
+                setFilteredPlugins(pluginData)
+            } catch (error) {
+                console.error('Error fetching plugin data', error)
+            } finally {
+                setLoading(false)
+            }
+        }
+        fetchPluginList()
+    }, [])
+
+    useEffect(() => {
+        const filtered = plugins.filter((plugin) => plugin.id.toLowerCase().includes(searchTerm.toLowerCase()))
+        setFilteredPlugins(filtered)
+    }, [searchTerm, plugins])
```

Comment:
I have a feeling this should be a `useMemo` instead of a `useEffect`: https://react.dev/reference/react/useMemo

Reply:
This is because in the body we are not dealing with side effects: https://www.w3schools.com/react/react_useeffect.asp

Reply:
I separated this logic into its own hook, which both gathers and filters the data, and I changed to useMemo.


```
import { useState, useEffect, useMemo } from 'react'
import axios from 'axios'
import { pluginList, IPluginData } from '../data/plugins'

const useSearchPlugins = (searchTerm: string) => {
    const [plugins, setPlugins] = useState<IPluginData[]>([])
    const [loading, setLoading] = useState<boolean>(true)

    useEffect(() => {
        const fetchPluginList = async () => {
            try {
                const pluginDataPromises = pluginList.map((plugin) => {
                    const url = `https://raw.githubusercontent.com/jenkins-infra/infra-statistics/gh-pages/plugin-installation-trend/${plugin.id}.stats.json`
                    return axios.get(url).then((response) => ({ ...plugin, chartData: response.data }))
                })
                console.log('Fetching plugin data...')
                const pluginData = await Promise.all(pluginDataPromises)
                setPlugins(pluginData)
            } catch (error) {
                console.error('Error fetching plugin data', error)
            } finally {
                setLoading(false)
            }
        }
        fetchPluginList()
    }, [])

    const filteredPlugins = useMemo(() => {
        return plugins.filter((plugin) => plugin.id.toLowerCase().includes(searchTerm.toLowerCase()))
    }, [searchTerm, plugins])

    return { filteredPlugins, loading }
}

export default useSearchPlugins

```

```
// PluginTrends.tsx
import { useState } from 'react'
import {
    Stack,
    Pagination,
    CircularProgress,
    TextField,
    Box,
    Grid,
    FormControl,
    InputLabel,
    Select,
    MenuItem,
} from '@mui/material'
import NavBar from '../../components/NavBar'
import useFetchAndFilterPlugins from '../../hooks/useFetchAndFilterPlugins'
import useSortPlugins from '../../hooks/useSortPlugins'
import usePagination from '../../hooks/usePagination'
import PluginCard from '../../components/PluginCard'

const PluginTrends: React.FC = () => {
    const [searchTerm, setSearchTerm] = useState<string>('')
    const { filteredPlugins, loading } = useFetchAndFilterPlugins(searchTerm)
    const { sortOption, setSortOption } = useSortPlugins(filteredPlugins)
    const itemsPerPage = 12

    const { page, handlePageChange, paginatedData, totalPages } = usePagination(filteredPlugins, itemsPerPage)

    return (
        <Stack
            sx={{
                backgroundColor: '#f0f0f0',
                minHeight: '100vh',
                minWidth: '100vw',
            }}
        >
            <NavBar />
            <Box
                sx={{
                    marginTop: '2.5rem',
                    display: 'flex',
                    flexDirection: 'row',
                    justifyContent: 'center',
                    alignItems: 'center',
                    gap: '2rem',
                    width: '100%',
                }}
            >
                <Box>
                    <TextField
                        label="Search Plugins"
                        variant="standard"
                        onChange={(e) => setSearchTerm(e.target.value)}
                    />
                </Box>
                <Box>
                    <FormControl variant="standard" sx={{ minWidth: 120 }}>
                        <InputLabel>Sort By</InputLabel>
                        <Select
                            value={sortOption}
                            onChange={(e) => setSortOption(e.target.value as 'alphabetical' | 'downloads')}
                            label="Sort By"
                        >
                            <MenuItem value="alphabetical">A - Z</MenuItem>
                            <MenuItem value="downloads">Installs High to Low</MenuItem>
                        </Select>
                    </FormControl>
                </Box>
            </Box>
            {loading ? (
                <CircularProgress />
            ) : (
                <>
                    <Grid
                        container
                        spacing={4}
                        sx={{
                            padding: '4rem',
                            paddingTop: '2rem',
                            marginTop: '0',
                        }}
                    >
                        {paginatedData.map((plugin) => (
                            <Grid item xs={12} sm={6} md={4} xl={3} key={plugin.id}>
                                <PluginCard plugin={plugin} />
                            </Grid>
                        ))}
                    </Grid>
                    <Pagination
                        count={totalPages}
                        page={page}
                        shape="rounded"
                        onChange={handlePageChange}
                        color="primary"
                        sx={{ marginTop: '2rem', marginBottom: '4rem', display: 'flex', justifyContent: 'center' }}
                    />
                </>
            )}
        </Stack>
    )
}

export default PluginTrends

```